clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6183,7 +6486,68 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "} // namespace detail\n";
   O << "} // namespace _V1\n";
   O << "} // namespace sycl\n";
-  O << "\n";
+
+  FreeFunctionFwdDeclEmitter FFFwdDeclEmitter(O, S.getLangOpts());
+  unsigned ShimCounter = 1;
+  int FreeFunctionCount = 0;
+  for (const KernelDesc &K : KernelDescs) {
+    if (!isFreeFunction(S, K.SyclKernel))
+      continue;
+
+    ++FreeFunctionCount;
+    O << "\n// Definition of " << K.Name << " as a free function kernel\n";
+    if (K.ParamTypes.size() > 0) {
+      for (const auto &P : K.ParamTypes) {
+        FFFwdDeclEmitter.Visit(P);
+      }
+    }
+    FFFwdDeclEmitter.VisitFunctionDecl(K.FreeFunctionCDecl);
```

Comment:
It doesn't look like you're actually doing any visitation. So I don't fully understand why you defined a type visitor for this.  Can't you just factor out the namespace logic and print this function directly?

Reply:
I've simplified the int header generation and don't need a separate FwdDeclEmitter.